File: app/src/main/java/com/codinglikeapirate/pocitaj/SolveActivity.java
package com.codinglikeapirate.pocitaj;

import android.content.Intent;
import android.graphics.drawable.AnimatedVectorDrawable;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.activity.EdgeToEdge;
import androidx.annotation.VisibleForTesting;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.graphics.Insets;
import androidx.core.view.ViewCompat;
import androidx.core.view.WindowInsetsCompat;

import com.codinglikeapirate.pocitaj.StrokeManager.DownloadedModelsChangedListener;
import com.codinglikeapirate.pocitaj.StrokeManager.ContentChangedListener;

import java.util.ArrayList;
import java.util.List;
import java.util.Set;

public class SolveActivity extends AppCompatActivity implements DownloadedModelsChangedListener, ContentChangedListener {

  private static final String TAG = "SolveActivity";

  @VisibleForTesting
  final StrokeManager strokeManager = new StrokeManager();

  private static final String modelLanguageTag = "en-US";
  private final ExerciseBook exerciseBook = new ExerciseBook();

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    EdgeToEdge.enable(this);
    setContentView(R.layout.activity_main);
    ViewCompat.setOnApplyWindowInsetsListener(findViewById(R.id.main), (v, insets) -> {
      Insets systemBars = insets.getInsets(WindowInsetsCompat.Type.systemBars());
      v.setPadding(systemBars.left, systemBars.top, systemBars.right, systemBars.bottom);
      return insets;
    });

    SolveView solveView = findViewById(R.id.solve_view);
    solveView.setExerciseBook(exerciseBook);
    solveView.setStrokeManager(strokeManager);

    TextView questionView = findViewById(R.id.question_view);
    solveView.setQuestionView(questionView);

    List<ImageView> progressIcons = new ArrayList<>();
    progressIcons.add(findViewById(R.id.progress_1));
    progressIcons.add(findViewById(R.id.progress_2));
    progressIcons.add(findViewById(R.id.progress_3));
    progressIcons.add(findViewById(R.id.progress_4));
    progressIcons.add(findViewById(R.id.progress_5));
    progressIcons.add(findViewById(R.id.progress_6));
    progressIcons.add(findViewById(R.id.progress_7));
    progressIcons.add(findViewById(R.id.progress_8));
    progressIcons.add(findViewById(R.id.progress_9));
    progressIcons.add(findViewById(R.id.progress_10));
    solveView.setProgressIcons(progressIcons);

    // strokeManager.deleteActiveModel();

    strokeManager.addContentChangedListener(solveView);
    strokeManager.addContentChangedListener(this);
    strokeManager.setDownloadedModelsChangedListener(this);
    strokeManager.setClearCurrentInkAfterRecognition(true);
    strokeManager.setTriggerRecognitionAfterInput(false);
    strokeManager.reset();
    strokeManager.setActiveModel(modelLanguageTag);
    strokeManager.download();
    strokeManager.refreshDownloadedModelsStatus();
    strokeManager.setTriggerRecognitionAfterInput(true);
  }

  public void onDownloadedModelsChanged(Set<String> downloadedLanguageTags) {
    for (String s : downloadedLanguageTags) {
      Log.i(TAG, "Downloaded models changed: " + s);
    }
  }

  @Override
  public void onNewRecognizedText(String text, boolean correct) {
    ImageView popupView = findViewById(R.id.popup_view);
    if (correct) {
      popupView.setImageResource(R.drawable.heart_animation);
    } else {
      popupView.setImageResource(R.drawable.teardrop);
    }
    popupView.setVisibility(View.VISIBLE);
    Drawable d = popupView.getDrawable();
    AnimatedVectorDrawable animation;
    if (d instanceof AnimatedVectorDrawable) {
      animation = (AnimatedVectorDrawable) d;
      animation.start();
    }

    new Handler(Looper.getMainLooper()).postDelayed(() -> {
      popupView.setVisibility(View.GONE);

      if (exerciseBook.getHistory().size() < 5)
        return;

      Intent intent = new Intent(this, ResultsActivity.class);
      Bundle bundle = new Bundle();

      int exerciseCount = exerciseBook.getHistory().size();
      String[] exercises = new String[exerciseCount];
      boolean[] corrects = new boolean[exerciseCount];
      boolean[] recognized = new boolean[exerciseCount];
      for (int i = 0; i < exerciseCount; i++) {
        exercises[i] = exerciseBook.getHistory().get(i).equation();
        corrects[i] = exerciseBook.getHistory().get(i).correct();
        recognized[i] = exerciseBook.getHistory().get(i).solved();
      }
      bundle.putStringArray("exercises", exercises);
      bundle.putBooleanArray("corrects", corrects);
      bundle.putBooleanArray("recognized", recognized);
      intent.putExtras(bundle);

      startActivity(intent, bundle);
    }, 500);
  }

  @Override
  public void onMisparsedRecognizedText(String text) {
    ImageView popupView = findViewById(R.id.popup_view);
    popupView.setImageResource(R.drawable.question);
    popupView.setVisibility(View.VISIBLE);
    Drawable d = popupView.getDrawable();
    AnimatedVectorDrawable animation;
    if (d instanceof AnimatedVectorDrawable) {
      animation = (AnimatedVectorDrawable) d;
      animation.start();
    }

    new Handler(Looper.getMainLooper()).postDelayed(() -> popupView.setVisibility(View.GONE), 500);
  }
}

File: app/src/main/java/com/codinglikeapirate/pocitaj/SolveActivity.kt
package com.codinglikeapirate.pocitaj

import android.content.Intent
import android.graphics.drawable.AnimatedVectorDrawable
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.View
import android.widget.ImageView
import android.widget.TextView
import androidx.annotation.VisibleForTesting
import androidx.appcompat.app.AppCompatActivity
import com.codinglikeapirate.pocitaj.StrokeManager.ContentChangedListener

/** Main activity which creates a StrokeManager and connects it to the DrawingView. */
class SolveActivity : AppCompatActivity(), StrokeManager.DownloadedModelsChangedListener,
    ContentChangedListener {
    @JvmField
    @VisibleForTesting
    val strokeManager = StrokeManager()
    private val modelLanguageTag = "en-US"
    private var exerciseBook = ExerciseBook()
    private val progressIcons = ArrayList<ImageView>()
    private lateinit var questionView : TextView
    private lateinit var solveView : SolveView

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        solveView = findViewById(R.id.solve_view)
        solveView.setExerciseBook(exerciseBook)
        solveView.setStrokeManager(strokeManager)
        updateExerciseBook(null, false)

        questionView = findViewById(R.id.question_view)

        progressIcons.add(findViewById(R.id.progress_1))
        progressIcons.add(findViewById(R.id.progress_2))
        progressIcons.add(findViewById(R.id.progress_3))
        progressIcons.add(findViewById(R.id.progress_4))
        progressIcons.add(findViewById(R.id.progress_5))
        progressIcons.add(findViewById(R.id.progress_6))
        progressIcons.add(findViewById(R.id.progress_7))
        progressIcons.add(findViewById(R.id.progress_8))
        progressIcons.add(findViewById(R.id.progress_9))
        progressIcons.add(findViewById(R.id.progress_10))
        updateViews()

        strokeManager.addContentChangedListener(this)
        strokeManager.addContentChangedListener(solveView)
        strokeManager.setDownloadedModelsChangedListener(this)
        strokeManager.setClearCurrentInkAfterRecognition(true)
        strokeManager.setTriggerRecognitionAfterInput(false)
        strokeManager.setActiveModel(modelLanguageTag)

        // add continuation here:
        strokeManager.download()

        strokeManager.refreshDownloadedModelsStatus()
        strokeManager.reset()
        strokeManager.setTriggerRecognitionAfterInput(true)

        val inkActivityButton = findViewById<View>(R.id.compose_ink)
        inkActivityButton.setOnClickListener {
            val intent = Intent(this, ExerciseBookActivity::class.java)
            startActivity(intent)
        }
    }

    private fun updateViews() {
        val history = exerciseBook.historyList
        progressIcons.forEachIndexed { i, icon ->
            icon.setImageResource(
                when {
                    history.size <= i -> R.drawable.cat_sleep
                    !history[i].solved() -> R.drawable.cat_big_eyes
                    !history[i].correct() -> R.drawable.cat_cry
                    else -> R.drawable.cat_heart
                }
            )
        }
        questionView.text = exerciseBook.last.question()
        solveView.invalidate()
    }
//
//    fun recognizeClick(v: View?) {
//        strokeManager.recognize()
//    }
//
//    fun clearClick(v: View?) {
//        strokeManager.reset()
//        val drawingView = findViewById<SolveView>(R.id.solve_view)
//        drawingView.clear()
//    }
//
//    fun deleteClick(v: View?) {
//        strokeManager.deleteActiveModel()
//    }

    override fun onDownloadedModelsChanged(downloadedLanguageTags: Set<String>) {
        for (s in downloadedLanguageTags) {
            Log.i(TAG, "Downloaded models changed: $s")
        }
    }

    private fun updateExerciseBook(text: String?, correct: Boolean) {
        if (exerciseBook.last.solved()) {
            Log.e(TAG, "last was solved")
        }
        val result = text?.toInt() ?: ExerciseBook.NOT_RECOGNIZED

        val correctlySolved = exerciseBook.last.solve(result)
        if (correctlySolved != correct) {
            Log.e(TAG, "Passed-through solution didn't match expected result")
        }
        Log.i(TAG, "Stats: ${exerciseBook.stats}")
        // do animation
        if (result != ExerciseBook.NOT_RECOGNIZED) {
            exerciseBook.generate()
        }
        strokeManager.expectedResult = exerciseBook.last.getExpectedResult()
    }

    override fun onNewRecognizedText(text: String?, correct: Boolean) {    val popupView = findViewById<ImageView>(R.id.popup_view)
        updateExerciseBook(text, correct)
        popupView.setImageResource(if (correct) R.drawable.heart_animation else R.drawable.teardrop)
        popupView.visibility = View.VISIBLE

        (popupView.drawable as? AnimatedVectorDrawable)?.start()

        Handler(Looper.getMainLooper()).postDelayed({
            popupView.visibility = View.GONE
            updateViews()

            if (exerciseBook.historyList.size < 5) {
                return@postDelayed
            }

            val intent = Intent(this, ResultsActivity::class.java)
            val bundle = Bundle()

            val exerciseCount = exerciseBook.historyList.size
            val exercises = Array(exerciseCount) { "" }
            val corrects = BooleanArray(exerciseCount)
            val recognized = BooleanArray(exerciseCount)

            for (i in 0 until exerciseCount) {
                val exercise = exerciseBook.historyList[i]
                exercises[i] = exercise.equation()
                corrects[i] = exercise.correct()
                recognized[i] = exercise.solved()
            }

            bundle.putStringArray(ResultsActivity.EXERCISES_KEY, exercises)
            bundle.putBooleanArray(ResultsActivity.CORRECTS_KEY, corrects)
            bundle.putBooleanArray(ResultsActivity.RECOGNIZED_KEY, recognized)

            intent.putExtras(bundle)
            startActivity(intent, bundle)
        }, 500)
    }

    override fun onMisparsedRecognizedText(text: String?) {
        val popupView = findViewById<ImageView>(R.id.popup_view)
        popupView.setImageResource(R.drawable.question)
        popupView.visibility = View.VISIBLE
        val d = popupView.getDrawable()
        if (d is AnimatedVectorDrawable) {
            d.start()
        }

        Handler(Looper.getMainLooper()).postDelayed({ popupView.visibility = View.GONE }, 500)
    }

    companion object {
        private const val TAG = "SolveActivity"
    }
}